src/conversation_engine/engine.py

```diff
@@ -308,102 +397,170 @@ def _handle_additional_info(self, text):
             rospy.loginfo("Action execution failed")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+
+            self._on_task_outcome_failed("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
         elif task_outcome.result == TaskOutcome.RESULT_UNKNOWN:
             rospy.loginfo("Action result: unknown")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+            self._on_task_outcome_unknown("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
         else:
             rospy.loginfo("Action result: other")
-            self._robot_to_user_pub.publish("".join(task_outcome.messages))
+            self._say_to_user("".join(task_outcome.messages))
 
             self._start_new_conversation()
 
     def _feedback_cb(self, feedback):
         self._latest_feedback = feedback
 
         rospy.loginfo(feedback.current_subtask)
-        self._robot_to_user_pub.publish(describe_current_subtask(feedback.current_subtask))
 
-    def _get_grammar_target(self, missing_field_path):
+    @staticmethod
+    def _get_grammar_target(missing_field_path):
+        """Determine which grammar target to use when a particular field of info is missing."""
         deepest_field_name = missing_field_path.split('.')[-1]
 
+        grammar_target = "T"
+
         if 'location' in deepest_field_name:
-            return 'ROOM_OR_LOCATION'
+            grammar_target = 'ROOM_OR_LOCATION'
+        elif 'entity' in deepest_field_name:
+            grammar_target = 'ROOM_OR_LOCATION'
+        elif 'target' in deepest_field_name:
+            grammar_target = 'NAMED_PERSON'
+        elif 'sentence' in deepest_field_name:
+            grammar_target = 'SAY_SENTENCE'
+        elif 'object' in deepest_field_name:
+            grammar_target = 'NAMED_OBJECT'
+
+        rospy.loginfo("Missing information '{}' must match '{}' in grammar".format(missing_field_path, grammar_target))
+        return grammar_target
+
+    @staticmethod
+    def _log_invalid_command(text):
+        with open("invalid_commands.txt", "a") as dump:
+            dump.writelines([text + "\n"])
+
+    def _say_to_user(self, message):
+        raise NotImplementedError("How to say something to the user must be implemented in the subclass. Signature: ```_say_to_user(message: str)```")
+
+    def _on_task_successful(self, message):
+        self._say_to_user(message)
+
+    def _on_request_missing_information(self, description, grammar, target):
+        rospy.loginfo("_request_missing_information('{}', '{}...', '{}')".format(description, grammar[:10], target))
+
+        if self.give_examples:
+            example = self._parser.get_random_sentence(self._state.target)
+            description += " For example: '{}'".format(example)
+        self._say_to_user(description)
+
+    def _on_task_outcome_failed(self, message):
```

Comment:
Why do we need all these single line methods? You can simply call `self._say_to_user`

Reply:
This is meant to be very extensible. For example, the ee-gpsr is now a subclass of this class and does some additional handling when a task is failed